File: DSA Practice/LeetCode/1721. Swapping Nodes in a Linked List/solution.java

//Definition for singly-linked list.
/*
public class ListNode {
     int val;
     ListNode next;
     ListNode() {}
     ListNode(int val) { this.val = val; }
     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 }
 */
 
class Solution {
    public ListNode swapNodes(ListNode head, int k) {
        int toSwap = 0;
        int swapWith = 0;
        int pointer = 0;
        int length = 0;

        ListNode temp = head;

        // get data for to swap
        while(head != null){
            pointer++;
            if (pointer == k){
                toSwap = head.val;
            }

            length++;
            head = head.next;
        }

        head = temp;

        if (length == 1){
            return head;
        }

        // toSwap
        for(int i = 0; i <= length - k - 1; i++){
            head = head.next;
        }

        swapWith = head.val;
        head.val = toSwap;
        head = temp;

        pointer = 0;
        // Swap
        while(head != null){
            pointer++;

            //System.out.println(head.val);
            if (pointer == k){
                head.val = swapWith;
                break;
            }

            head = head.next;
        }

        /*
        System.out.println(toSwap);
        System.out.println(length);
        System.out.println(swapWith);
        */

        head = temp;

        return head;
    }
}

File: DSA Practice/LeetCode/1721. Swapping Nodes in a Linked List/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapNodes(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        length = 0
        first = 0
        second = 0
        pointer = 0
        temp = head

        while(head):
            pointer += 1

            if (pointer == k):
                first = head.val

            length += 1
            head = head.next

        head = temp

        for _ in range(length - k):
            head = head.next
        
        second = head.val
        head.val = first
        head = temp
        
        pointer = 0
        while(head):
            pointer += 1

            if (pointer == k):
                head.val = second
                break

            head = head.next
        
        head = temp
        return head